### ISSUE TITLE
`Float` values with integer value beyond `int` lose precision if bound to `long`

### ISSUE THREAD

**Issue by maroux** [FasterXML/jackson-databind#1940]
Jackson version: 2.9.4

Sample code demonstrating the problem:
```
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;

@JsonTypeInfo(use=JsonTypeInfo.Id.CLASS, include=JsonTypeInfo.As.PROPERTY, property="_class")
public class LongContainer {
    @JsonProperty
    private Long longObj;

    public static void main(String[] args) throws IOException {
        ObjectMapper objectMapper = new ObjectMapper();
        JsonNode tree = objectMapper.readTree("{\"longObj\": 1519348261000.0, \"_class\": \"LongContainer\"}");
        LongContainer obj = objectMapper.treeToValue(tree, LongContainer.class);
        System.out.println(obj.longObj);
    }
}
```

It's caused by [this line](https://github.com/FasterXML/jackson-databind/commit/6a1152cd446230b192b7d5b056cc1cb3877be684#diff-5c0023d081050275137dd9b6afc1ece9R1692). Curiously enough, it works fine if you deserialize a plain float value into `Long.class` since that uses `TreeTraversingParser` instead of `TokenBuffer`.

**Comment by cowtowncoder** [comment]
Wow. That is curious line. Must have been cut'n pasted. Thanks, will need to fix.

**Comment by cowtowncoder** [comment]
Thank you for reporting this -- seems to be a regression due to refactoring, not covered (yet) with unit tests. Added test, fixed, will be in `2.9.5`.

**Comment by maroux** [comment]
Awesome. Thanks for the quick fix!
